channeldb/invoices.go

```diff
@@ -1241,79 +1260,79 @@ func (d *DB) updateInvoice(hash lntypes.Hash, invoices, settleIndex *bbolt.Bucke
 		return &invoice, err
 	}
 
-	// Update invoice state.
-	invoice.State = update.State
+	// If there is nothing to update, return early.
+	if update == nil {
+		return &invoice, nil
+	}
 
-	now := d.now()
+	now := d.Now()
 
-	// Update htlc set.
-	for key, htlcUpdate := range update.Htlcs {
-		htlc, ok := invoice.Htlcs[key]
+	// Update invoice state if the update descriptor indicates an invoice
+	// state change.
+	if update.State != nil {
+		err := updateInvoiceState(&invoice, hash, *update.State)
+		if err != nil {
+			return nil, err
+		}
 
-		// No update means the htlc needs to be canceled.
-		if htlcUpdate == nil {
-			if !ok {
-				return nil, fmt.Errorf("unknown htlc %v", key)
-			}
-			if htlc.State != HtlcStateAccepted {
-				return nil, fmt.Errorf("can only cancel " +
-					"accepted htlcs")
+		if update.State.NewState == ContractSettled {
+			err := setSettleMetaFields(
+				settleIndex, invoiceNum, &invoice, now,
+			)
+			if err != nil {
+				return nil, err
 			}
-
-			htlc.State = HtlcStateCanceled
-			htlc.ResolveTime = now
-			invoice.AmtPaid -= htlc.Amt
-
-			continue
 		}
+	}
 
-		// Add new htlc paying to the invoice.
-		if ok {
-			return nil, fmt.Errorf("htlc %v already exists", key)
+	// Process add actions from update descriptor.
+	for key, htlcUpdate := range update.AddHtlcs {
```

Comment:
what's the rationale behind doing the invoice state update before adding/canceling HTLCs? (which is a change of order)

Reply:
This isn't changing is this commit, or is it? It only pulls apart the add and cancel actions?

Reply:
Or maybe you mean the next commit? But there we still update the state first, as we do in master?

Reply:
Yes, the rationale is that we don't need to loop through the htlcs twice. So the invoice state is leading and that is set first. Then we go through the htlcs and update their states based on the invoice state and what we got from the update descriptor